_source/_posts/2019-12-20-five-tools-improve-java.adoc

```diff
@@ -0,0 +1,345 @@
+
+def it = jApiClasses.iterator()
+while (it.hasNext()) {
+    def jApiClass = it.next()
+
+    if (jApiClass.getChangeStatus() != UNCHANGED) {
+        def methodIt = jApiClass.getMethods().iterator()
+        while (methodIt.hasNext()) {
+            def method = methodIt.next()
+            def methodChanges = method.getCompatibilityChanges()
+            methodChanges.remove(METHOD_NEW_DEFAULT)
+        }
+    }
+}
+return jApiClasses
+----
+
+Take a look at the https://siom79.github.io/japicmp/[japicmp project documentation] for more examples.
+
+== 5. Don't Skip Code Reviews
+
+Using the above tools can help make your code reviews more effective. The goal should be to automate as much as possible out of your code review so that the human element can shine through. Have you ever asked, "is there a test for this" in a code review? Automate that, send your coverage data to a tool like https://codecov.io/[Codecov], which can add the coverage delta to your pull requests. If your project has strict code style guidelines, you can use https://checkstyle.org/[Checkstyle].
+
+Code reviews are great; they provide an opportunity for both the author and the reviewer to learn from each other and ask questions, suggest alternatives, or discuss other architectural topics. The reviewer shouldn't be wasting time checking for things that a program can detect.
+
+TIP: **Pro Tip:** Keep the number of changes in your reviews small and to the point. Nobody wants to review hundreds of potentially unrelated code changes at once.
+
+== Bonus: Scan your Dependencies for Vulnerabilities
+
+Your code is just a small percentage of your overall application. Dependencies (direct and indirect) make up the rest. Keeping on top of vulnerabilities in those dependencies is not something you can do manually. Luckily for us, there are several tools available to help us out.
+
+- https://www.owasp.org/index.php/OWASP_Dependency_Check[OWASP Dependency Check] - I've been using the Maven plugin with success for years. The only downside is there is a high rate of false-positive matches that requires updating an "exclusion" file in your repository.
+- https://snyk.io/[Snyk.io] - Offers dependency scanning, and includes additional security issues that are not official in the https://nvd.nist.gov/[NIST National Vulnerability Database]
+- https://dependabot.com/[GitHub Dependabot] - GitHub has been rolling out Dependabot, and there is a good chance it's already scanning your public repositories. I've had mixed success in the past, specifically when it comes to Maven multi-module projects. I'm sure this will improve in the future.
+- Many more! Have a favorite dependency scanner; let us know in the comments!
+
+WARNING: One key thing to remember the code in your repository doesn't always match the code that is running in production. Make sure you track the dependencies in your production code too!
+
+== Want to Build More Secure Applications? Learn More!
```

Comment:
This header is kinda funny. When you mouse over it, the link indicator makes it jump to two lines. Maybe change it to "Want More..."?